lib/serviceDiscovery/helpers.js

```diff
@@ -8,35 +8,95 @@ const interfaces = os.networkInterfaces();
 const ifaceNames = Object.keys(interfaces);
 
 /**
- * Builds a list of every announce-able IP on the server
+ * Chech if the list is empty
  *
- * @returns {string} The ip list
- * @todo Support from env variables?
+ * @param {*} addresses
+ * @param {*} errorMessage
+ */
+function assertNotEmpty(addresses, errorMessage) {
+	if (addresses.length === 0) {
+		throw Error(errorMessage);
+	}
+}
+
+/**
+ * Add all valid IP to an existing list of IP
+ *
+ * @param {*} addresses
+ * @param {*} announced
+ * @param {*} useInternalAddresses
+ */
+function pushIps(addresses, announced, useInternalAddresses) {
+	if (!useInternalAddresses) {
+		addresses = addresses.filter((address) => address.internal === false);
+	}
+
+	addresses = addresses
+		.filter((address) => address.family === 'IPv4')
+		.map((address) => address.address);
+
+	return announced.concat(addresses);
+}
+
+/**
+ * Return all valid IP for a single interface
+ *
+ * @param {*} ifaceName
  */
-exports.getAnnounceIps = memoizee(function () {
-	const announceIps = [];
+function getAnnouncedIpsForInterface(ifaceName) {
+	const addresses = interfaces[ifaceName];
 
-	for (var i = 0; i < ifaceNames.length; i++) {
-		var addresses = interfaces[ifaceNames[i]];
+	if (!addresses) {
+		throw new Error('Interface ' + ifaceName + ' could not be found');
+	}
 
-		for (var j = 0; j < addresses.length; j++) {
-			var address = addresses[j];
+	const announced = pushIps(addresses, [], true);
+	assertNotEmpty(announced, 'Interface ' + ifaceName + ' does not define any IP addresses');
 
-			// skip anything that doesn't interest us
-			if (address.internal) {
-				continue;
-			}
+	return announced;
+}
 
-			announceIps.push(address.address);
-		}
+/**
+ * Builds a list of IP. If ifaceName is not set, we return
+ * all public IP of the host (essentially, excluding localhost).
+ *
+ * @returns {string} The ip list
+ */
+exports.getAnnounceIps = memoizee(function (ifaceName) {
+	// If the interface is specified, we may want to specify localhost
+	if (ifaceName) {
+		return getAnnouncedIpsForInterface(ifaceName);
 	}
 
-	return announceIps.sort();
+	const announced = ifaceNames.reduce((announced, name) => pushIps(interfaces[name], announced, false), []);
+	assertNotEmpty(announced, 'Could not find any valid IP addresses');
+
+	// Note: the list of IP **MUST** stay in the order received by
+	// `os.networkInterfaces()`; otherwise, running MAGE on a Docker Swarm
+	// will break in some cases.
```

Comment:
And yet the order cannot be guaranteed by virtue of how JavaScript works (ie: we're lucky V8 keeps it in a certain order, but even Node cannot guarantee that will remain the case).

Could we not filter with a configurable mask for example?

Reply:
The order of the interfaces cannot, but the array of IP for each interface can.

Reply:
Ah, and those were the problem? Got it 👍